server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -0,0 +1,77 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDocumentSession, ISession } from "@fluidframework/server-services-client";
+import { IDocument, ICollection } from "@fluidframework/server-services-core";
+import { BaseTelemetryProperties, Lumberjack } from "@fluidframework/server-services-telemetry";
+
+export async function getSession(documentId: string,
+    ordererUrl: string,
+    historianUrl: string,
+    tenantId: string,
+    documentsCollection: ICollection<IDocument>): Promise<IDocumentSession> {
+    const lumberjackProperties = {
+        [BaseTelemetryProperties.tenantId]: tenantId,
+        [BaseTelemetryProperties.documentId]: documentId,
+    };
+
+    const tempDocument: IDocument = await documentsCollection.findOne({ documentId });
+    let tempSession: ISession = tempDocument.session;
+    const deli = JSON.parse(tempDocument.deli);
+    const isScribeEmpty = tempDocument.scribe === "";
+    const scribe = isScribeEmpty ? "" : JSON.parse(tempDocument.scribe);
+    let hasSessionLocationChanged: boolean = false;
+    const isSessionAlive: boolean = tempSession !== undefined ? tempSession.isSessionAlive : true;
+    if (tempSession === undefined) {
+        tempSession = {
+            ordererUrl,
+            historianUrl,
+            isSessionAlive: true,
+        };
+        await documentsCollection.upsert({
+            documentId,
+        }, {
+            deli: JSON.stringify(deli),
+            scribe: isScribeEmpty ? "" : JSON.stringify(scribe),
+            session: tempSession,
+        }, {
+        });
+        const ms: string = `The Session ${JSON.stringify(tempSession)} was inserted into the document collection`;
+        Lumberjack.info(ms, lumberjackProperties);
+    }
+
+    if (!tempSession.isSessionAlive) {
+        // Reset logOffset, ordererUrl, and historianUrl when switching cluster.
+        if (tempSession.ordererUrl !== ordererUrl) {
```

Comment:
Is there a chance that `ordererUrl` be empty or undefined? Meaning it might not be the case that the `ordererUrl` changed, but instead that the caller passed an invalid `ordererUrl` parameter?

Reply:
The order url and historian url should come from server in config.
If they are invalid, the whole service shouldn't work. :)